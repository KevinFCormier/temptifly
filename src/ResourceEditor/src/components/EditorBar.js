'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { SearchInput } from '@patternfly/react-core'
import '../css/editor-bar.scss'
import {
  CloseIcon,
  UndoIcon,
  RedoIcon,
  NextIcon,
  PreviousIcon,
} from './Icons'
import { defaultResourceEditorOptions } from '../ResourceEditor'

class EditorButton extends React.Component {
  static propTypes = {
    button: PropTypes.object,
    command: PropTypes.string,
    handleClick: PropTypes.func
  };

  handleClick = () => {
    const { command, button: { disabled } } = this.props
    if (!disabled) {
      document.activeElement.blur()
      this.props.handleClick(command)
    }
  };

  handleKeyPress = e => {
    if (e.key === 'Enter') {
      this.props.handleClick(this.props.command)
    }
  };

  renderIcon = icon => {
    switch(icon) {
    case 'close':
      return <CloseIcon />
    case 'undo':
      return <UndoIcon />
    case 'redo':
      return <RedoIcon />
    case 'next':
      return <NextIcon />
    case 'previous':
      return <PreviousIcon />
    }
  };

  render() {
    const { button: { disabled, tooltip, icon, spacer, command } } = this.props
    if (spacer) {
      return <div className="editor-bar-spacer" />
    } else {
      const classes = classNames({
        'editor-bar-button': true,
        [`${command}`]: true,
        disabled
      })
      return (
        <div
          className={classes}
          tabIndex={0}
          role={'button'}
          aria-label={tooltip}
          title={tooltip}
          onClick={this.handleClick}
          onKeyPress={this.handleKeyPress}
        >
          {icon ?
            this.renderIcon(icon)
            : (
              <div>{tooltip}</div>
            )}
        </div>
      )
    }
  }
}

class EditorBar extends React.Component {
  static propTypes = {
    handleEditorCommand: PropTypes.func,
    handleSearchChange: PropTypes.func,
    hasRedo: PropTypes.bool,
    hasUndo: PropTypes.bool,
    title: PropTypes.string,
    type: PropTypes.string
  };

  constructor(props) {
    super(props)
    this.state = {
      searchName: ''
    }
  }

  handleClick = command => {
    this.props.handleEditorCommand(command)
  };

  handleSearch = (searchName) => {
    this.props.handleSearchChange(searchName)
    this.setState({ searchName })
  };

  handleClear = () => {
    this.props.handleSearchChange('')
    this.setState({ searchName: '' })
  };

  render() {
    const { hasUndo, hasRedo, type, title } = this.props
    const { searchName } = this.state
    
    const undoButtons = [
      {
        command: 'undo',
        tooltip: defaultResourceEditorOptions.undoTip,
        icon: 'undo',
        disabled: !hasUndo
      },
      {
        command: 'redo',
        tooltip: defaultResourceEditorOptions.redoTip,
        icon: 'redo',
        disabled: !hasRedo
      }
    ]

    const nextButtons = [
      {
        command: 'previous',
        tooltip: defaultResourceEditorOptions.previousTip,
        icon: 'previous',
        disabled: !searchName
      },
      {
        command: 'next',
        tooltip: defaultResourceEditorOptions.nextTip,
        icon: 'next',
        disabled: !searchName
      }
    ]

    const resetButtons = [
      {
        command: 'restore',
        tooltip: defaultResourceEditorOptions.resetTip,
        disabled: !hasUndo && !hasRedo
      }
    ]

    const closeButtons = [
      {
        command: 'close',
        tooltip: defaultResourceEditorOptions.closeTip,
        icon: 'close'
      }
    ]

    const searchTitle = defaultResourceEditorOptions.findTip
    return (
      <div className="editor-bar">
        <div className="editor-bar-group">
          <div className="editor-bar-title">{title}</div>
        </div>
        <div className="editor-bar-group">
          <div className="editor-bar-toolbar">
            <div className="editor-bar-section">
              {resetButtons.map(button => {
                const { command } = button
                return (
                  <EditorButton
                    key={command}
                    command={command}
                    button={button}
                    handleClick={this.handleClick}
                  />
                )
              })}
            </div>
            <div className="editor-bar-section">
              {undoButtons.map(button => {
                const { command } = button
                return (
                  <EditorButton
                    key={command}
                    command={command}
                    button={button}
                    handleClick={this.handleClick}
                  />
                )
              })}
            </div>
            <div className="editor-bar-section">
              <div
                className="editor-bar-search"
                role="region"
                aria-label={searchTitle}
                id={searchTitle}
              >
                <SearchInput
                  id={`template-editor-search-${type}`}
                  value={searchName}
                  aria-label={searchTitle}
                  placeholder={searchTitle}
                  onChange={this.handleSearch}
                  onClear={this.handleClear}
                />
                {nextButtons.map(button => {
                  const { command } = button
                  return (
                    <EditorButton
                      key={command}
                      command={command}
                      button={button}
                      handleClick={this.handleClick}
                    />
                  )
                })}
              </div>
            </div>
          </div>
        </div>
        <div className="editor-bar-close">
          <div className="editor-bar-section">
            {closeButtons.map(button => {
              const { command } = button
              return (
                <EditorButton
                  key={command}
                  command={command}
                  button={button}
                  handleClick={this.handleClick}
                />
              )
            })}
          </div>
        </div>
      </div>
    )
  }
}

export default EditorBar
