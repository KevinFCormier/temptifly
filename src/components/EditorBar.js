'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Search } from 'carbon-components-react'
import '../scss/editor-bar.scss'
import '../../graphics/diagramIcons.svg'

class EditorButton extends React.Component {
  static propTypes = {
    button: PropTypes.object,
    command: PropTypes.string,
    handleClick: PropTypes.func
  };

  handleClick = () => {
    const { command, button: { disabled } } = this.props
    if (!disabled) {
      document.activeElement.blur()
      this.props.handleClick(command)
    }
  };

  handleKeyPress = e => {
    if (e.key === 'Enter') {
      this.props.handleClick(this.props.command)
    }
  };

  render() {
    const { button: { disabled, tooltip, icon, spacer, command } } = this.props
    if (spacer) {
      return <div className="editor-bar-spacer" />
    } else {
      const classes = classNames({
        'editor-bar-button': true,
        [`${command}`]: true,
        disabled
      })
      return (
        <div
          className={classes}
          tabIndex={0}
          role={'button'}
          aria-label={tooltip}
          title={tooltip}
          onClick={this.handleClick}
          onKeyPress={this.handleKeyPress}
        >
          {icon ? (
            <svg>
              <use href={`#diagramIcons_${icon}`} />
            </svg>
          ) : (
            <div>{tooltip}</div>
          )}
        </div>
      )
    }
  }
}

class EditorBar extends React.Component {
  static propTypes = {
    handleEditorCommand: PropTypes.func,
    handleSearchChange: PropTypes.func,
    hasRedo: PropTypes.bool,
    hasUndo: PropTypes.bool,
    i18n: PropTypes.func,
    title: PropTypes.string,
    type: PropTypes.string
  };

  constructor(props) {
    super(props)
    this.state = {
      searchName: ''
    }
  }

  handleClick = command => {
    this.props.handleEditorCommand(command)
  };

  handleSearch = ({ target }) => {
    const searchName = target.value || ''
    this.props.handleSearchChange(searchName)
    this.setState({ searchName })
  };

  render() {
    const { hasUndo, hasRedo, type, title, i18n } = this.props
    const { searchName } = this.state

    const undoButtons = [
      {
        command: 'undo',
        tooltip: i18n('editor.bar.undo'),
        icon: 'undo',
        disabled: !hasUndo
      },
      {
        command: 'redo',
        tooltip: i18n('editor.bar.redo'),
        icon: 'redo',
        disabled: !hasRedo
      }
    ]

    const nextButtons = [
      {
        command: 'previous',
        tooltip: i18n('editor.bar.previous'),
        icon: 'previous',
        disabled: !searchName
      },
      {
        command: 'next',
        tooltip: i18n('editor.bar.next'),
        icon: 'next',
        disabled: !searchName
      }
    ]

    const resetButtons = [
      {
        command: 'restore',
        tooltip: i18n('editor.bar.reset'),
        disabled: !hasUndo && !hasRedo
      }
    ]

    const closeButtons = [
      {
        command: 'close',
        tooltip: i18n('editor.bar.close'),
        icon: 'close'
      }
    ]

    const searchTitle = i18n('find.label')
    return (
      <div className="editor-bar">
        <div className="editor-bar-group">
          <div className="editor-bar-title">{title}</div>
        </div>
        <div className="editor-bar-group">
          <div className="editor-bar-toolbar">
            <div className="editor-bar-section">
              {resetButtons.map(button => {
                const { command } = button
                return (
                  <EditorButton
                    key={command}
                    command={command}
                    button={button}
                    handleClick={this.handleClick}
                  />
                )
              })}
            </div>
            <div className="editor-bar-section">
              {undoButtons.map(button => {
                const { command } = button
                return (
                  <EditorButton
                    key={command}
                    command={command}
                    button={button}
                    handleClick={this.handleClick}
                  />
                )
              })}
            </div>
            <div className="editor-bar-section">
              <div
                className="editor-bar-search"
                role="region"
                aria-label={searchTitle}
                id={searchTitle}
              >
                <Search
                  id={`template-editor-search-${type}`}
                  labelText=""
                  aria-label={searchTitle}
                  placeHolderText={searchTitle}
                  small={true}
                  onChange={this.handleSearch}
                />
                {nextButtons.map(button => {
                  const { command } = button
                  return (
                    <EditorButton
                      key={command}
                      command={command}
                      button={button}
                      handleClick={this.handleClick}
                    />
                  )
                })}
              </div>
            </div>
          </div>
        </div>
        <div className="editor-bar-close">
          <div className="editor-bar-section">
            {closeButtons.map(button => {
              const { command } = button
              return (
                <EditorButton
                  key={command}
                  command={command}
                  button={button}
                  handleClick={this.handleClick}
                />
              )
            })}
          </div>
        </div>
      </div>
    )
  }
}

export default EditorBar
