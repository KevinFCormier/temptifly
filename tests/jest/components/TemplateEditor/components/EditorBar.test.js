'use strict'

import React from 'react'
import EditorBar from '../../../../../src/components/EditorBar'
import renderer from 'react-test-renderer'

describe('EditorBar component', () => {
  it('renders as expected', () => {
    const fn = jest.fn()
    const exceptions = [{ text: 'bad', row: 0 }]
    const component = renderer.create(
      <EditorBar
        hasUndo={false}
        hasRedo={true}
        exceptions={exceptions}
        gotoEditorLine={fn}
        handleEditorCommand={fn}
        handleSearchChange={fn}
        i18n={fn}
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
