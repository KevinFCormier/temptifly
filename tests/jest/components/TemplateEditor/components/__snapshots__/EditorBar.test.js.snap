// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorBar component renders as expected 1`] = `
<div
  className="editor-bar"
>
  <div
    className="editor-bar-group"
  >
    <div
      className="editor-bar-title"
    />
  </div>
  <div
    className="editor-bar-group"
  >
    <div
      className="editor-bar-toolbar"
    >
      <div
        className="editor-bar-section"
      >
        <div
          aria-label={undefined}
          className="editor-bar-button restore"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
          title={undefined}
        >
          <div />
        </div>
      </div>
      <div
        className="editor-bar-section"
      >
        <div
          aria-label={undefined}
          className="editor-bar-button undo disabled"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
          title={undefined}
        >
          <svg
            height="16px"
            role="presentation"
            version="1.1"
            viewBox="0 0 18 12"
            width="16px"
          >
            <g
              fill="none"
              fillRule="evenodd"
              stroke="none"
              strokeWidth="1"
            >
              <g
                fill="#EDEDED"
                transform="translate(-827.000000, -218.000000)"
              >
                <g
                  transform="translate(827.000000, 218.000000)"
                >
                  <path
                    d="M14,5.3 C15.5,5.3 16.8,6.6 16.8,8 C16.8,9.5 15.5,10.7 14,10.7 L10,10.7 L10,12 L14,12 C16.3,12 18,10.3 18,8 C18,5.8 16.2,4 14,4 L2.4,4 L4.9,1.5 L4,0.6 L0,4.6 L4,8.6 L4.9,7.7 L2.4,5.3 L14,5.3 Z"
                    id="Fill"
                  />
                </g>
              </g>
            </g>
          </svg>
        </div>
        <div
          aria-label={undefined}
          className="editor-bar-button redo"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
          title={undefined}
        >
          <svg
            height="16px"
            role="presentation"
            version="1.1"
            viewBox="0 0 18 12"
            width="16px"
          >
            <g
              fill="none"
              fillRule="evenodd"
              stroke="none"
              strokeWidth="1"
            >
              <g
                fill="#EDEDED"
                transform="translate(-857.000000, -218.000000)"
              >
                <g
                  transform="translate(857.000000, 218.000000)"
                >
                  <path
                    d="M4,5.3 C2.5,5.3 1.2,6.6 1.2,8 C1.2,9.5 2.5,10.7 4,10.7 L8,10.7 L8,12 L4,12 C1.7,12 0,10.3 0,8 C0,5.8 1.8,4 4,4 L15.6,4 L13.1,1.5 L14,0.6 L18,4.6 L14,8.6 L13.1,7.7 L15.6,5.2 L4,5.2 L4,5.3 Z"
                    id="Fill"
                  />
                </g>
              </g>
            </g>
          </svg>
        </div>
      </div>
      <div
        className="editor-bar-section"
      >
        <div
          aria-label={undefined}
          className="editor-bar-search"
          id={undefined}
          role="region"
        >
          <div
            aria-labelledby="template-editor-search-undefined-label"
            className="bx--search bx--search--sm"
            role="search"
          >
            <svg
              alt=""
              aria-label=""
              className="bx--search-magnifier"
              fill={undefined}
              fillRule="evenodd"
              height="16"
              name={undefined}
              role="img"
              style={undefined}
              viewBox="0 0 16 16"
              width="16"
            >
              <title>
                
              </title>
              <path
                d="M6.5 12a5.5 5.5 0 1 0 0-11 5.5 5.5 0 0 0 0 11zm4.936-1.27l4.563 4.557-.707.708-4.563-4.558a6.5 6.5 0 1 1 .707-.707z"
              />
            </svg>
            <label
              className="bx--label"
              htmlFor="template-editor-search-undefined"
              id="template-editor-search-undefined-label"
            >
              
            </label>
            <input
              aria-label={undefined}
              className="bx--search-input"
              id="template-editor-search-undefined"
              onChange={[Function]}
              placeholder=""
              type="text"
            />
            <button
              aria-label={undefined}
              className="bx--search-close bx--search-close--hidden"
              onClick={[Function]}
              type="button"
            >
              <svg
                alt="Provide a description that will be used as the title"
                aria-label="Provide a description that will be used as the title"
                className={undefined}
                fill={undefined}
                fillRule="evenodd"
                height="16"
                name={undefined}
                role="img"
                style={undefined}
                viewBox="0 0 16 16"
                width="16"
              >
                <title>
                  Provide a description that will be used as the title
                </title>
                <path
                  d="M8 6.586L5.879 4.464 4.464 5.88 6.586 8l-2.122 2.121 1.415 1.415L8 9.414l2.121 2.122 1.415-1.415L9.414 8l2.122-2.121-1.415-1.415L8 6.586zM8 16A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"
                />
              </svg>
            </button>
          </div>
          <div
            aria-label={undefined}
            className="editor-bar-button previous disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title={undefined}
          >
            <svg
              height="16px"
              role="presentation"
              version="1.1"
              viewBox="0 0 18 18"
              width="16px"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeWidth="1"
              >
                <g
                  fill="#EDEDED"
                  transform="translate(-1225.000000, -215.000000)"
                >
                  <path
                    d="M1234,216.5 C1238.14214,216.5 1241.5,219.857864 1241.5,224 C1241.5,228.142136 1238.14214,231.5 1234,231.5 C1229.85786,231.5 1226.5,228.142136 1226.5,224 C1226.5,219.857864 1229.85786,216.5 1234,216.5 L1234,216.5 Z M1234,215.25 C1229.16751,215.25 1225.25,219.167508 1225.25,224 C1225.25,228.832492 1229.16751,232.75 1234,232.75 C1238.83249,232.75 1242.75,228.832492 1242.75,224 C1242.75,219.167508 1238.83249,215.25 1234,215.25 Z M1230.35502,224.334055 L1235.91465,224.334055 L1233.00667,227.242032 L1233.89056,228.125915 L1238.30997,223.706498 L1233.89498,219.2915 L1233.00225,220.166545 L1235.91907,223.08336 L1230.35944,223.08336 L1230.35502,224.334055 L1230.35502,224.334055 Z"
                    id="Fill"
                    transform="translate(1234.000000, 224.000000) scale(-1, 1) translate(-1234.000000, -224.000000) "
                  />
                </g>
              </g>
            </svg>
          </div>
          <div
            aria-label={undefined}
            className="editor-bar-button next disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
            title={undefined}
          >
            <svg
              height="16px"
              role="presentation"
              version="1.1"
              viewBox="0 0 18 18"
              width="16px"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeWidth="1"
              >
                <g
                  fill="#EDEDED"
                  transform="translate(-1255.000000, -215.000000)"
                >
                  <path
                    d="M1264,216.5 C1268.14214,216.5 1271.5,219.857864 1271.5,224 C1271.5,228.142136 1268.14214,231.5 1264,231.5 C1259.85786,231.5 1256.5,228.142136 1256.5,224 C1256.5,219.857864 1259.85786,216.5 1264,216.5 L1264,216.5 Z M1264,215.25 C1259.16751,215.25 1255.25,219.167508 1255.25,224 C1255.25,228.832492 1259.16751,232.75 1264,232.75 C1268.83249,232.75 1272.75,228.832492 1272.75,224 C1272.75,219.167508 1268.83249,215.25 1264,215.25 Z M1260.35502,224.334055 L1265.91465,224.334055 L1263.00667,227.242032 L1263.89056,228.125915 L1268.30997,223.706498 L1263.89498,219.2915 L1263.00225,220.166545 L1265.91907,223.08336 L1260.35944,223.08336 L1260.35502,224.334055 L1260.35502,224.334055 Z"
                    id="Fill"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="editor-bar-close"
  >
    <div
      className="editor-bar-section"
    >
      <div
        aria-label={undefined}
        className="editor-bar-button close"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
        title={undefined}
      >
        <svg
          height="16px"
          role="presentation"
          version="1.1"
          viewBox="-4 -4 16 16"
          width="16px"
        >
          <g>
            <path
              d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
